### PR TITLE
Stop freezing main thread

### PR DESCRIPTION
**Pull request by Ksengine**
stop freezing main thread
- winforms Edge(supported, tested)
- winforms CEF (supported, not tested)
- winforms IE (supported, not tested)
- gtk(supported, tested)
- qt(supported, tested on linux only)
- cocoa(not supported)

so,
- test IE
- test CEF
- test qt on mac
- develop cocoa

other developers are welcomed

### CODE REVIEW

**Comment by r0x0r** [comment]
@Ksengine can you add an example and a simple test case?

**Comment by Ksengine** [comment]
sure

**Comment by Ksengine** [comment]
something is wrong in AppVeyor build

**Comment by Ksengine** [comment]
added example
added test
added docs
added example to docs

File: webview/platforms/qt.py [hunk 1]
```diff
@@ -611,8 +643,13 @@ def _create():
...
         global _app
         _app = QApplication.instance() or QApplication([])
-
         _create()
+        
+        if _multiprocessing:
+            p = Process()
+            p.exec_ = _app.exec_
+            return p
+
```
Comment (r0x0r) [inline webview/platforms/qt.py:652]: What is the rationale behind a custom process class? Why not use python's multiprocessing as with GTK?

Comment (Ksengine) [reply webview/platforms/qt.py:652]: ?

Comment (Ksengine) [reply webview/platforms/qt.py:652]: > Why not use python's multiprocessing as with GTK?

in `platforms.gtk.py` file i added it
this is qt
qt runs even without gui loop(app.exec_)
why we run gui loop?
i created a custom(and simple) `Process` class

File: webview/platforms/qt.py [hunk 2]
```diff
@@ -593,6 +593,38 @@ def _get_debug_port():
...
+    authkey = b''
+    sentinel = None
+    
+    def join(self, timeout=None):
+        if timeout:
+            from time import sleep
+            sleep(timeout)
+            self.kill()
+        else:
+            self.exec_()
```
Comment (Ksengine) [inline webview/platforms/qt.py:612]: if `join` called without `timeout` arg it starts gui loop

File: webview/platforms/qt.py [hunk 3]
```diff
@@ -593,6 +593,38 @@ def _get_debug_port():
...
+            
+    def is_alive(self):
+        return self.alive
+    
+    def kill(self):
```
Comment (Ksengine) [inline webview/platforms/qt.py:617]: destroy all windows(like killing gui loop)

**Review by Ksengine — COMMENTED** [review]
minor edits

**Comment by r0x0r** [comment]
@Ksengine let's revisit the API proposal to make sure we are on the same page.

1) In a non-blocking mode `webview.start` would return a process object. 
2) Process is a common object (`process.py` or something like that) that is shared between all gui frameworks. `join` is the only needed function I think. At least there is no need to duplicate functionality of `window` like close.
3) The main thread would exit killing all the windows once the code after `start` finishes executing unless `join()` is invoked or any other user-defined blocking mechanism is invoked (e.g. a http server)
4) multiprocessing architecture can likely be implemented as a common multiprocessing implementation inside `__init__.py`. I have not investigated non-blocking capabilities of each framework, but if there are no restrictions on running GUI on a child process, this would be the most straight-forward implementation.

Does this make sense to you?

**Comment by Ksengine** [comment]
## 1
ok i will explain it.
### In a non-blocking mode `webview.start` returns a process object.**(i did it)**
chosen gui library returns process to start and start returns it
all gui libs can't use `multiprocessing.Process`
- we should use `fork` method to work it properly
    _according to the python docs:_
    > `fork` : **Available on Unix only. The default on Unix.**

- some gui libs dont want multiprocessing
    - Qt - no need to run gui loop,so it returns **fake Process**.also `multiprocessing.Process` raises Exceptions on MacOS(explained above)
    - Winforms(CEF, Edge, IE) - we can use `System.Threading.Thread` from .net library. wrapped it using custom `Process`(.net like API to python like)

**Comment by Ksengine** [comment]
## 2
 process is a  `multiprocessing.Process` like object.
custom `Process` is used in some gui libs as i explained above
`Process` has common methods visit [python docs](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process)
python developers are familiar with it.
so i added all methods of `multiprocessing.Process` to custom `Process`
custom `Process` is used only in Qt and Winforms.
If you want i can move it to `process.py` (but its only used in qt and winforms)

**Comment by Ksengine** [comment]
# 3. 
use `daemon`
i had a project called `DicksonUI`
It uses pywebview.
I want to keep all windows until user close them.
if someone want feature like this?

**Comment by Ksengine** [comment]
## 4.
as i explained above, we had to use various methods in various platforms
if you can find method to use  `multiprocessing` in `fork` mode on any platform, we can include it in `__init__.py` file
i had'nt found solution

**Comment by Ksengine** [comment]
also some other questions

**Comment by Ksengine** [comment]
with multiprocessing
1. cocoa not works(i have'nt mac computer)
2. gtk is working on Linux only

other contributors should fix cocoa
if yo can help me with GTK

**Comment by r0x0r** [comment]
I am fine with multiprocessing as well as native non-blocking gui, as long as the API is the same across platforms. Ie `p = start(block=False)` shall always return a common process object. There is no need to replicate other methods of Python's Process, `join()` is enough for now.

**Comment by r0x0r** [comment]
Cocoa seems to work with multiprocessing.

**Comment by Ksengine** [comment]
> join() is enough for now.
`multiprocessing.Process` return many properties and methods other than `join`
to make the API same across platforms, I had added other methods to fake Process objects.
so always `start(block=False` returns `multiprocessing.Process` like object.

**Comment by Ksengine** [comment]
> Cocoa seems to work with multiprocessing.

**in `fork` method?**

-------------------------
can you run below scripts and give me the logs on macos
maybe we can add support to macos cocoa

```python
import multiprocessing as mp 
print(mp.get_context())
print(mp.get_context('fork'))
```
```python
import multiprocessing as mp 
print(mp.set_start_method('fork'))
```

**Comment by r0x0r** [comment]
It works with a spawn. With fork it displays an empty window and throws a warning. I suspect that WKWebView has problems with forking.

**Comment by r0x0r** [comment]
> `multiprocessing.Process` return many properties and methods other than `join`
> 
> to make the API same across platforms, I had added other methods to fake Process objects.
> so always `start(block=False` returns `multiprocessing.Process` like object.

You are thinking it the other way around. Instead of replicating `multiprocessing.Process` functionality, a better approach is to always return a custom Process object that would have only `join`. Less is more.  This way API would be uniform across the platforms no matter whether multiprocessing is used or not.

File: webview/platforms/cocoa.py [hunk 4]
```diff
@@ -433,6 +434,11 @@ def first_show(self):
             self._add_view_menu()
 
             BrowserView.app.activateIgnoringOtherApps_(Foundation.YES)
+            if _multiprocessing:
```
Comment (Ksengine) [inline webview/platforms/cocoa.py:437]: i have no MacOS computer, so you and other contributors should add multiprocessing to macos.
use this lines to add multiprocessing to macos cocoa

**Review by Ksengine — COMMENTED** [review]
macos support

**Comment by Ksengine** [comment]
> You are thinking it the other way around. Instead of replicating multiprocessing.Process functionality, a better approach is to always return a custom Process object that would have only join. Less is more. This way API would be uniform across the platforms no matter whether multiprocessing is used or not.

you can force users to use `join` only

**Comment by Ksengine** [comment]
can you accept this pull now?

File: webview/platforms/qt.py (hunk 2 above)
Comment (r0x0r) [reply webview/platforms/qt.py:612]: Extract this class to `process.py`. All the platforms should return the same object. join() is the only needed function for now. If timeout is needed kill can be left as well.

File: webview/platforms/gtk.py [hunk 5]
```diff
@@ -260,7 +260,11 @@ def show(self):
         if gtk.main_level() == 0:
             if self.pywebview_window.hidden:
                 self.window.hide()
-            gtk.main()
+            if _multiprocessing:
+                from multiprocessing import Process
+                p = Process(target=gtk.main)
+                p.start()
+                return p
```
Comment (r0x0r) [inline webview/platforms/gtk.py:267]: Return a custom process object

File: webview/platforms/winforms.py [hunk 6]
```diff
@@ -646,6 +646,28 @@ def subprocess_args(include_stdout=True):
...
+    def is_alive(self):
+        return self.alive
+    
+    def terminate(self):
+        self.kill()
+        
+    def close(self):
+        self.kill()
+
+
```
Comment (r0x0r) [inline webview/platforms/winforms.py:670]: Use a common process object

File: webview/platforms/winforms.py [hunk 7]
```diff
@@ -676,7 +698,15 @@ def create():
...
-        thread.Join()
+        if not _multiprocessing:
+            thread.Join()
+        else:
+            p = Process()
+            p.join = thread.Join
+            p.is_alive = thread.IsAlive
+            p.name = thread.Name
+            p.kill = thread.Abort
+            return thread
```
Comment (r0x0r) [inline webview/platforms/winforms.py:709]: Return a common process object.

**Review by r0x0r — CHANGES_REQUESTED** [review]

File: webview/platforms/winforms.py (hunk 6 above)
Comment (Ksengine) [reply webview/platforms/winforms.py:670]: no
custom Process is used in qt and winforms
winforms api and qt api is different
but this processes are same

File: webview/platforms/qt.py (hunk 2 above)
Comment (Ksengine) [reply webview/platforms/qt.py:612]: are you sure?
developers want other methods and properties(specially `close`)
other maybe other methods and properties wanted by someone.
I want `close` method

File: docs/examples/noblock.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    
+    process = webview.start(block = False)
+    
+    # third window created after gui loop started
+    third_window = webview.create_window('Window #3', html='<h1>Third Window</h1>')
+    process.join()
+
+```
+
+`process` returned by `webview.start` is [multiprocessing.Process](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process) like object
```
[the remark this answers is not in the document]
Comment (Ksengine) [reply docs/examples/noblock.md:30]: only `join` and `close` methods are documented in docs

File: webview/platforms/winforms.py (hunk 6 above)
Comment (r0x0r) [reply webview/platforms/winforms.py:670]: Winforms and QT APIs are different but they can be unified via a common process object. This is what I am asking.

**Comment by Ksengine** [comment]
no comments?

**Comment by r0x0r** [comment]
My comment regarding process object still stands. I won't accept this pull request, until process is implemented as described.

**Comment by Ksengine** [comment]
I did those changes
this is docs
>`process` returned by `webview.start` is the process which runs GUI event loop.
> ## `Process` object
> ### `join(timeout=None)`
> join GUI event loop to main thread
> > Warning: This will freeze main thread.
> - timeout(optional) - timeout in seconds(int). default is None. if timeout, kills GUI event loop after timeout.
> ### `is_alive()`
> returns True if GUI event loop is running, otherwise False (GUI event loop killed)
> ### `close()`
> close all windows and kill GUI event loop.
where can I include this docs

**Comment by r0x0r** [comment]
Thanks.
What happened to the GTK implementation? It can be done in a similar way as Winforms by mapping multiprocessing process properties  to our Process.

Can you give me edit rights to this PR. I will push cocoa implementation.

**Comment by Ksengine** [comment]
I'll do it

**Comment by Ksengine** [comment]
> I will push cocoa implementation.

amazing!!!
I added a tick to `Allow edits by maintainers`. its help text says
> If checked, users with write access to **r0x0r/pywebview** can add new commits to your master branch. You can always change this setting later.

so you have edit rights.

**Comment by Ksengine** [comment]
explain the method you found when you pushed cocoa implementation.

**Comment by Ksengine** [comment]
> I will push cocoa implementation.

?

**Comment by r0x0r** [comment]
@Ksengine 
I could not get it to work after all. The problem is pickling webview objects when they are passed to multiprocessing's Process. I have tried also `dill` for pickling but it choked on pyobjc objects. Same problem applies to GTK.

The only way it works is when defining a target function at the top level ie 

```
import webview
import multiprocessing


def create_window():
    window = webview.create_window('Simple browser', 'https://pywebview.flowrl.com/hello')
    webview.start()

if __name__ == '__main__':
    # Create a standard webview window
    p = multiprocessing.Process(target=create_window)
    p.start()

    print('w00t')
    p.join()
```

**Comment by Ksengine** [comment]
Yes I know

**Comment by Ksengine** [comment]
In `cocoa.py`
Create function to run gui loop
Pass windows as array to function
Run function using multiprocessing

**Comment by Ksengine** [comment]
Freezing apps works with multiprocessing simply

**Comment by Ksengine** [comment]
So we can use multiprocessing in freezed apps.
For python interpreter
1. check if cocoa can run without GUI loop(like Qt)
2. create a function to run gui loop, pass browser windows array as argument and run it in new process
3. run python interpreter using subprocess and run whole gui on it(this should work.but bad and hard way)

Check 1 and 2
Then we can talk about 3

**Comment by r0x0r** [comment]
Regarding the example I posted above. Considering you can achieve a non-blocking mode in all platforms using a couple of code lines, what use is in implementing a "native" blocking mode?

**Comment by Ksengine** [comment]
multiprocessing is heavy. Qt and Winforms has easy ways
its simpler than your example.
how can we call window methods. sometimes it will be buggy. so we should use tricks like shared memory, pipes, managers, queues , locks etc. 
on Windows and MacOS you cant use global variables in your function.

**Comment by Ksengine** [comment]
> So we can use multiprocessing in freezed apps.
> For python interpreter
> 
> 1. check if cocoa can run without GUI loop(like Qt)
> 2. create a function to run gui loop, pass browser windows array as argument and run it in new process
> 3. run python interpreter using subprocess and run whole gui on it(this should work.but bad and hard way)
> 
> Check 1 and 2
> Then we can talk about 3

?

**Comment by r0x0r** [comment]
> So we can use multiprocessing in freezed apps.
> For python interpreter
> 
> 1. check if cocoa can run without GUI loop(like Qt)
> 2. create a function to run gui loop, pass browser windows array as argument and run it in new process
> 3. run python interpreter using subprocess and run whole gui on it(this should work.but bad and hard way)
> 
> Check 1 and 2
> Then we can talk about 3

Sorry I don't follow. What problem exactly are you trying to solve?

I won't put any more effort into getting Cocoa/GTK to work. Somebody else can take a look at this. Multiprocessing example would do the job for now. I will document this approach.

**Review by Ksengine — COMMENTED** [review]
test MacOs for noblock

**Comment by r0x0r** [comment]
freeze_support does not make a difference. fork is not supported on macOS and spawn fails to pickle webview objects.

**Comment by Ksengine** [comment]
> freeze_support does not make a difference
freeze_support is for freezing applications

**Comment by Ksengine** [comment]
> test MacOs for noblock

In my last change, -
It passes all globals to child Process.
please test it because I have no MacOS machine

**Comment by Ksengine** [comment]
even no appveyor build on MacOS?

**Comment by r0x0r** [comment]
I have tested it. It crashes with forking and throws a pickle exception with spawning.

**Comment by Ksengine** [comment]
sorry, I cant fix it without Mac computer

**Comment by Ksengine** [comment]
> I have tested it. It crashes with forking and throws a pickle exception with spawning.

after my last  review?

**Comment by Ksengine** [comment]
> > I have tested it. It crashes with forking and throws a pickle exception with spawning.
> 
> after my last review?

?

**Comment by Ksengine** [comment]
no answer from you

**Comment by analogaio** [comment]
Hope you guys end up making this work! I have been slightly following along, one thing I did notice is that when I open the webview from a Process, it has an extremely long startup time that increases exponentially as I open more webviews.

**Comment by Ksengine** [comment]
> Hope you guys end up making this work! I have been slightly following along, one thing I did notice is that when I open the webview from a Process, it has an extremely long startup time that increases exponentially as I open more webviews.

looks like a bug?

**Comment by Ksengine** [comment]
@analogaio please tell more info
- operation system - windows, macos or linux?
- used guilib - Qt, GTK, IE(internet exploref), Edge(mahtml), CEF(chromium embedded framework), Cocoa(macos)
- your sample code
etc

**Comment by Ksengine** [comment]
> I have been slightly following along

Thnak you @analogaio

**Comment by Ksengine** [comment]
> Hope you guys end up making this work!

mostly done. but MacOS cocoa platform not supported yet
I have not MacOS machine. If you have, please test last updated code.

you made a new Issue, please comment to #627 . we can discuss with @r0x0r

**Comment by Ksengine** [comment]
# I found a solution to MacOS
threre is also a bug in GTK. This solution is good for both
- new `gui_proc` method
```python
def gui_proc(*args, **kwargs):
    import webview
    webview.create_window(*args)
    webview.start(**kwargs)
```
- `create_window` method in `cocoa.py`  or `gtk.py`
  if multiprocessing enabled,
  1. it extract the properties of `window` object
  2. it extract the variables of webview_module(only wanted ones to pass to `webview.start`)

  run `gui_proc` on spawn process with  as args and 2 as kwargs

**Comment by Ksengine** [comment]
its better if we can add 2 new files to `platforms` folder

**Comment by Ksengine** [comment]
> # I found a solution to MacOS
> threre is also a bug in GTK. This solution is good for both
> 
> * new `gui_proc` method
> 
> ```python
> def gui_proc(*args, **kwargs):
>     import webview
>     webview.create_window(*args)
>     webview.start(**kwargs)
> ```
> 
> * `create_window` method in `cocoa.py`  or `gtk.py`
>   if multiprocessing enabled,
>   
>   1. it extract the properties of `window` object
>   2. it extract the variables of webview_module(only wanted ones to pass to `webview.start`)
>   
>   run `gui_proc` on spawn process with  as args and 2 as kwargs

:question:

**Comment by r0x0r** [comment]
@Ksengine if you find a way to incorporate this the codebase without rehauling the whole architecture, that would be great. Otherwise this code segment looks like something to be done by user.

**Comment by Ksengine** [comment]
I will try

**Comment by Ksengine** [comment]
please run following test on macos
```python
import webview
import multiprocessing


def create_window():
    window = webview.create_window('Simple browser', 'https://pywebview.flowrl.com/hello')
    webview.start()

if __name__ == '__main__':
    # Create a standard webview window
    p = multiprocessing.Process(target=create_window)
    p.start()
    p2 = multiprocessing.Process(target=create_window)
    p2.start()
``

**Comment by Ksengine** [comment]
# GTK bug
when gtk used as guilib
after `webview.start` called we cant create new windows.

I can fix GTK and Macos using `spawn` method in multiprocessing.
but they will not work when frozen because `spawn` not supported. (only `fork` supported)
**can you find a way to use `spawn` when frozen.**

**Comment by Ksengine** [comment]
can you find a way to use `spawn` when frozen.

**Comment by Ksengine** [comment]
## Question

can I replace some semaphores in gtk.py with `multiprocessing.Queue`
**like :**
https://github.com/r0x0r/pywebview/blob/addc7352b7fdd35d1264485bdf8ddb6ac2d5ac6f/webview/platforms/gtk.py#L538

**Comment by Ksengine** [comment]
## found a solution for GTK
but its very hard and workaround.

**Review by Ksengine — COMMENTED** [review]
GTK finished

**Comment by Ksengine** [comment]
Its hard to add multiprocessing to this  library. because of design architecture.
please motivate developers to make their contributions multiprocessing friendly.

**Comment by Ksengine** [comment]
sorry its not working.
I nedd help.

**Comment by Ksengine** [comment]
tests successful
but we should place this code on separate file.
we can do same for cocoa.

**Comment by Ksengine** [comment]
hi @r0x0r 
I did it.
test MacOS

**Comment by Ksengine** [comment]
@r0x0r I am waiting for a reply

**Comment by r0x0r** [comment]
Exception has occurred: ImportError
cannot import name '_multiprocessing'
  File "/Users/roman/Code/pywebview/webview/guilib.py", line 5, in <module>
    from webview import _multiprocessing
  File "/Users/roman/Code/pywebview/webview/__init__.py", line 22, in <module>
    from webview.guilib import initialize
  File "/Users/roman/Code/pywebview/examples/simple_browser.py", line 1, in <module>
    import webview

**Comment by Ksengine** [comment]
Its impossible

**Comment by Ksengine** [comment]
its ok

**Comment by Ksengine** [comment]
test again

**Comment by r0x0r** [comment]
Exception has occurred: ImportError
cannot import name '_multiprocessing'
  File "/Users/roman/Code/pywebview/webview/guilib.py", line 5, in <module>
    from webview import _multiprocessing
  File "/Users/roman/Code/pywebview/webview/__init__.py", line 22, in <module>
    from webview.guilib import initialize
  File "/Users/roman/Code/pywebview/examples/simple_browser.py", line 1, in <module>
    import webview

**Comment by Ksengine** [comment]
@r0x0r test it on macos again.

**Comment by r0x0r** [comment]
```
[pywebview] Using Cocoa
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/roman/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/lib/python/debugpy/__main__.py", line 45, in <module>
    cli.main()
  File "/Users/roman/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/Users/roman/.vscode/extensions/ms-python.python-2020.11.371526539/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 267, in run_file
    runpy.run_path(options.target, run_name=compat.force_str("__main__"))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/roman/Code/pywebview/examples/simple_browser.py", line 10, in <module>
    webview.start(block=False)
  File "/Users/roman/Code/pywebview/webview/__init__.py", line 108, in start
    return guilib.create_window(windows[0])
  File "/Users/roman/Code/pywebview/webview/platforms/mp_cocoa.py", line 74, in create_window
    windows[window.uid] = window
TypeError: list indices must be integers or slices, not str
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/roman/Code/pywebview/webview/platforms/mp_cocoa.py", line 64, in event_loop
    uid, event = event_queue.get()
  File "<string>", line 2, in get
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/managers.py", line 757, in _callmethod
    kind, result = conn.recv()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError
```

Sorry I dont have time to test every change. I won't give this any more attention.

**Comment by Ksengine** [comment]
I have no MacOS computer. So you should do tests or you can make a fix for cocoa.
If you cant, then I can clean macos part.
Then we can left it to a future developer and close or merge this pr.

**Comment by Ksengine** [comment]
@r0x0r It's ok. MacOS is working and tested(on github actions).

**Comment by Ksengine** [comment]
Hi @r0x0r can you merge this.

**Comment by r0x0r** [comment]
macOS no workie 
```

[pywebview] Using Cocoa
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/roman/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/__main__.py", line 45, in <module>
    cli.main()
  File "/Users/roman/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 444, in main
    run()
  File "/Users/roman/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/roman/Code/pywebview/examples/noblock.py", line 11, in <module>
    webview.start(block=False)
  File "/Users/roman/Code/pywebview/webview/__init__.py", line 127, in start
    return guilib.create_window(windows[0])
  File "/Users/roman/Code/pywebview/webview/platforms/mp_cocoa.py", line 74, in create_window
    windows[window.uid] = window
TypeError: list indices must be integers or slices, not str
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/roman/Code/pywebview/webview/platforms/mp_cocoa.py", line 64, in event_loop
    uid, event = event_queue.get()
  File "<string>", line 2, in get
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/managers.py", line 757, in _callmethod
    kind, result = conn.recv()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError

```

**Comment by Ksengine** [comment]
Not stale

**Comment by Ksengine** [comment]
> This PR has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

not stale

**Comment by Ksengine** [comment]
> This PR has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

Not stale

**Comment by Ksengine** [comment]
> This PR has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

Not stale

**Pull request closed without merging**